coderd/workspaceapps/proxy.go

```diff
@@ -361,41 +362,84 @@ func (s *Server) HandleSubdomain(middlewares ...func(http.Handler) http.Handler)
 				return
 			}
 
-			if !s.handleAPIKeySmuggling(rw, r, AccessMethodSubdomain) {
-				return
-			}
-
-			token, ok := ResolveRequest(rw, r, ResolveRequestOptions{
-				Logger:              s.Logger,
-				SignedTokenProvider: s.SignedTokenProvider,
-				DashboardURL:        s.DashboardURL,
-				PathAppBaseURL:      s.AccessURL,
-				AppHostname:         s.Hostname,
-				AppRequest: Request{
-					AccessMethod:      AccessMethodSubdomain,
-					BasePath:          "/",
-					UsernameOrID:      app.Username,
-					WorkspaceNameOrID: app.WorkspaceName,
-					AgentNameOrID:     app.AgentName,
-					AppSlugOrPort:     app.AppSlugOrPort,
-				},
-				AppPath:  r.URL.Path,
-				AppQuery: r.URL.RawQuery,
-			})
-			if !ok {
-				return
+			// REVIEW: Like mentioned in coderd.go maybe we should extract the app
+			// using middleware that way we can do this in a single top-level CORS
+			// handler?  Or just do the URL parsing twice.
+			var corsmw func(next http.Handler) http.Handler
+			origin := r.Header.Get("Origin")
+			if originApp, ok := s.parseOrigin(origin); ok && originApp.Username == app.Username {
+				corsmw = cors.Handler(cors.Options{
+					AllowedOrigins: []string{origin},
+					AllowedMethods: []string{
+						http.MethodHead,
+						http.MethodGet,
+						http.MethodPost,
+						http.MethodPut,
+						http.MethodPatch,
+						http.MethodDelete,
+					},
+					AllowedHeaders:   []string{"*"},
+					AllowCredentials: true,
+				})
+			} else {
+				corsmw = cors.Handler(cors.Options{
+					AllowedOrigins:   []string{""}, // The middleware defaults to *.
+					AllowedMethods:   []string{},
+					AllowedHeaders:   []string{},
+					AllowCredentials: false,
+				})
```

Comment:
Put this logic into the `AllowOriginFunc` field instead of this conditional MW

Reply:
So much better!